### PR TITLE
refactor(launch): fetch git repos with protocol in remote url

### PR DESCRIPTION
**Pull request by bcsherma**
Description
-----------

- Fixes WB-16089

Git job are associated with 2 pieces of information about the source code: a remote, and a commit. When jobs are created automatically from a run, we always store the remote uri of a single repository and a commit hash (not a branch name or other ref).

The remote uri either looks like `git@host.com:user/repo.git` or `https://host.com/user/repo`

The other mechanism for creating a git job is the `wandb job create git` CLI command. This command stores a git uri and commit exactly as provided by the user, but will fail if the remote is invalid. The commit passed by the user is not validated and is stored exactly as entered in the job.

Another behavior is that we always parse the remote uri into components and then use an https url that we reconstruct to fetch the repo. This means users can't create jobs that will use ssh creds for the clone.

Our reference handling includes a lot of sophisticated parsing that allows us to handle urls that include file paths, like
```
https://github.com/wandb/simple-launch-job/blob/branch/main.py
```
I think that we were originally hoping to support creating jobs from a single file or subdir in a git repo. However, I don't this flow is working as intended on main. Any automatically created job still includes the entire repo, and while `wandb job create git https://github.com/wandb/simple-launch-job/blob/branch/main.py` works it will clone the whole repository and include it in the git repo. I think for now our users have good ways of specifying an entrypoint or subdir for a job - they can use `--entrypoint` or `--dockerfile` to make the effective root of the job whatever they want.

The remote urls that we need to handle can only come from our sdk, which just gives a repo url, or wandb job create, where we can make it a rule that you just provide a repo url and we validate it before creating a job.

This PR makes the following changes:
- We no longer parse the remote uris. We just try to use them and complain if there is failure.
- We don't support `https://github.com/wandb/simple-launch-job/blob/branch/main.py` - this just won't work, this command wants a repo url now! 

review questions
------------------

I'm particularly interested in feedback on:
- Do we actually want to keep the parsing features this PR removes?
- Do we think it makes sense to clone with url and protocol of the remote url?

Copilot
------------

### Summary
🐛🛠️🔄

This pull request refactors and improves the code related to launching jobs from GitHub repositories. It simplifies the `GitHubReference` class and the `_create_job` and `_fetch_git_repo` functions, adds better error handling and validation, and supports SSH URLs. It also fixes and updates the corresponding tests in `test_job.py` and deletes the redundant `test_github_reference.py` file.

> _To create jobs from git repos with ease_
> _We refactored the code and fixed some bugs_
> _We used `GitHubReference` to parse the URLs_
> _And handled errors with tracebacks and asserts_
> _Now the function works well with SSH and HTTPS_

### Walkthrough
*  Simplify and update the GitHubReference class and its usage ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-0ca82984d2ee112939c1832334437c6b28e1c2ae8ef77954626d905f629b9765L44-R55), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-0ca82984d2ee112939c1832334437c6b28e1c2ae8ef77954626d905f629b9765L160-R85), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-0ca82984d2ee112939c1832334437c6b28e1c2ae8ef77954626d905f629b9765L226-R107), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-cbbb677decdd1d0e95eb6d34e11c2ac3878d178ff6793291e2dd36dc9788fabcL480-R485), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L275-R281), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L282-R288), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L289-R295))
  * Replace the parse method with a constructor that takes a remote and a ref as arguments ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-0ca82984d2ee112939c1832334437c6b28e1c2ae8ef77954626d905f629b9765L44-R55), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-cbbb677decdd1d0e95eb6d34e11c2ac3878d178ff6793291e2dd36dc9788fabcL480-R485), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L275-R281))
  * Support SSH URLs for the remote argument ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-0ca82984d2ee112939c1832334437c6b28e1c2ae8ef77954626d905f629b9765L44-R55))
  * Remove unnecessary attributes and methods, such as directory, file, default_branch, and _update_path ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-0ca82984d2ee112939c1832334437c6b28e1c2ae8ef77954626d905f629b9765L44-R55), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-0ca82984d2ee112939c1832334437c6b28e1c2ae8ef77954626d905f629b9765L226-R107), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L289-R295))
  * Set the path attribute to the working directory of the repository, and the commit_hash attribute to the hexsha of the checked out commit ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-0ca82984d2ee112939c1832334437c6b28e1c2ae8ef77954626d905f629b9765L160-R85), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L282-R288))
  * Handle git command errors and raise a LaunchError with a helpful message ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-0ca82984d2ee112939c1832334437c6b28e1c2ae8ef77954626d905f629b9765L160-R85))
* Add error handling and validation to the _create_job function in `create_job.py` ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5R6), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L112-R126), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L310-R315))
  * Import the traceback module to format the exception traceback ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5R6))
  * Add a try-except block around the _make_metadata_for_partial_job function call, and log any exception as a termerror with the traceback ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L112-R126))
  * Add an assertion that the entrypoint is not None, as it is required for creating a job from a git source ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L310-R315))
* Simplify the mock_fetch_repo function and the test_create_git_job function in `test_job.py` ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-36536a3b4e0515cc355034a9b275677498c90367681ee743495ff94e11ee9962L131-R142), [link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-36536a3b4e0515cc355034a9b275677498c90367681ee743495ff94e11ee9962L169-R169))
  * Remove the unnecessary "commit/path/" prefix from the files that are created in the temporary directory ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-36536a3b4e0515cc355034a9b275677498c90367681ee743495ff94e11ee9962L131-R142))
  * Update the assertion to match the new entrypoint value that does not include the "commit/path/" prefix ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-36536a3b4e0515cc355034a9b275677498c90367681ee743495ff94e11ee9962L169-R169))
* Remove the redundant "root" key from the metadata dictionary returned by the _create_repo_metadata function in `create_job.py` ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L362))
  * The "root" key was used to store the ref.repo attribute, which is the name of the git repository, but it is not used anywhere else in the code, and it is redundant with the "uri" key, which stores the full remote URL of the git repository ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-d4afe212673314c0a8a04be48ee33f5c8448b42fb33dcc25cf949950168fc5d5L362))
* Update the docstring of the _fetch_git_repo function in `utils.py` ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-cbbb677decdd1d0e95eb6d34e11c2ac3878d178ff6793291e2dd36dc9788fabcL469-R472))
  * Change the return type annotation from str to Optional[str], as the function can return None if the version argument is None and the GitHubReference object does not have a ref or a default branch ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-cbbb677decdd1d0e95eb6d34e11c2ac3878d178ff6793291e2dd36dc9788fabcL469-R472))
  * Remove the part that says the function defaults to the head commit of the master branch if version is unspecified, as this is not true anymore ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-cbbb677decdd1d0e95eb6d34e11c2ac3878d178ff6793291e2dd36dc9788fabcL469-R472))
* Delete the `test_github_reference.py` file, as it is not needed anymore ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-465689b7a08e03190bfc908fc3d9a5ce26dea43055dae6ab9aa695b9c3cfc614))
  * The file contained tests for the deprecated parse method of the GitHubReference class, which is replaced by the constructor ([link](https://github.com/wandb/wandb/pull/6545/files?diff=unified&w=0#diff-465689b7a08e03190bfc908fc3d9a5ce26dea43055dae6ab9aa695b9c3cfc614))

### CODE REVIEW

File: wandb/sdk/launch/create_job.py [hunk 1]
```diff
@@ -109,15 +110,20 @@ def _create_job(
...
+            git_hash=git_hash,
+            runtime=runtime,
+            path=path,
+            entrypoint=entrypoint,
+        )
+        if not metadata:
+            return None, "", []
+    except Exception as e:
+        e_traceback_str = "".join(traceback.format_tb(e.__traceback__))
+        wandb.termerror(f"Error creating job: {e}/n{e_traceback_str} ")
```
Comment (nickpenaranda) [inline wandb/sdk/launch/create_job.py:126]: typo? is that supposed to be a newline?

File: wandb/sdk/launch/github_reference.py [hunk 2]
```diff
@@ -41,114 +39,19 @@ def _parse_netloc(netloc: str) -> Tuple[Optional[str], Optional[str], str]:
...
-        assert self.host
-        auth = self.username or ""
-        if self.password:
-            auth += f":{self.password}"
-        if auth:
-            auth += "@"
-        return f"{PREFIX_HTTPS}{auth}{self.host}"
+        Arguments:
+            remote: A remote URL or URI
+            ref: A branch or commit
```
Comment (nickpenaranda) [inline wandb/sdk/launch/github_reference.py:47]: will this work for tags too?

**Review by nickpenaranda — APPROVED** [review]
couple small comments

File: wandb/sdk/launch/github_reference.py [hunk 3]
```diff
@@ -3,9 +3,7 @@
 import re
```
Comment (KyleGoyette) [inline wandb/sdk/launch/github_reference.py:3]: should we rename this to git_reference?

File: wandb/sdk/launch/create_job.py (hunk 1 above)
Comment (bcsherma) [reply wandb/sdk/launch/create_job.py:126]: Yeah I'm actually going to remove this traceback it's just for capturing git errors and adds a lot of noise

File: wandb/sdk/launch/github_reference.py (hunk 2 above)
Comment (bcsherma) [reply wandb/sdk/launch/github_reference.py:47]: Yes, I'm updating the comment

**Review by gtarpenning — APPROVED** [review]
tested and ✅ but we should improve the error messaging here next

**Pull request merged**